### PR TITLE
Improving clarity demo

### PR DESCRIPTION
**Pull request by maksimdrachov**
todo:

- [x] Add support for mermaid graphs
  - [mgaitan/sphinxcontrib-mermaid](https://github.com/mgaitan/sphinxcontrib-mermaid)
 
See also: [maksimdrachov/opencyphal-tutorial](https://github.com/maksimdrachov/opencyphal-tutorial)

### CODE REVIEW

**Comment by maksimdrachov** [comment]
I would like to add the following [mermaid](https://mermaid.js.org/syntax/flowchart.html) diagrams to improve clarity:


```mermaid
---
title: Legend
---
flowchart TB
    id1[/Message-publisher\]
    id2[\Message-subscriber/]
    id3{{Service}}
    id4[subject_or_service_id:type]
```

```mermaid
---
title: demo_app.py
---
flowchart TB
    subgraph 42:org.opencyphal.pycyphal.demo.demo_app
        direction TB
        subgraph heartbeat_publisher
            direction TB
            heartbeat_publisher_1[/uavcan.node.Heartbeat.1.0\]
        end
        heartbeat_publisher --> uavcan.node.heartbeat
        subgraph temperature_setpoint
            direction TB
            temperature_setpoint_1[\uavcan.si.unit.temperature.Scalar_1/]
        end
        2345:uavcan.si.sample.temperature.Scalar --> temperature_setpoint
        subgraph temperature_measurement
            direction TB
            temperature_measurement_1[\uavcan.si.sample.temperature.Scalar_1/]
            
        end
        2346:uavcan.si.unit.voltage.Scalar --> temperature_measurement
        subgraph heater_voltage
            direction TB
            heater_voltage_1[/uavcan.si.unit.voltage.Scalar_1\]
        end
        heater_voltage --> 2347:uavcan.si.unit.voltage.Scalar
        subgraph least_squares
            direction TB
            least_squares_1{{arasaka_corp.PerformLinearLeastSquaresFit_1}}
        end
        123:arasaka_corp.PerformLinearLeastSquaresFit_1.Request --> least_squares
        least_squares --> 123:arasaka_corp.PerformLinearLeastSquaresFit_1.Response
    end
```

```mermaid
---
title: plant.py
---
flowchart TB
    subgraph 43:org.opencyphal.pycyphal.demo.plant
        direction TB
        subgraph heartbeat_publisher
            direction TB
            heartbeat_publisher_1[/uavcan.node.Heartbeat.1.0\]
        end
        heartbeat_publisher --> uavcan.node.heartbeat
        subgraph temperature
            direction TB
            temperature_publisher_1[/uavcan.si.sample.temperature.Scalar_1\]
        end
        temperature --> 2346:uavcan.si.sample.temperature.Scalar
        subgraph voltage
            direction TB
            voltage_1[\uavcan.si.unit.voltage.Scalar_1/]
            
        end
        2347:uavcan.si.unit.voltage.Scalar --> voltage
    end
```

File: docs/pages/demo.rst [hunk 1]
```diff
@@ -187,7 +187,7 @@ launch the following in a new terminal and leave it running (``y`` is a convenie
     export UAVCAN__UDP__IFACE=127.0.0.1
     y sub --with-metadata uavcan.node.heartbeat uavcan.diagnostic.record    # You should see heartbeats
 
-Now let's see how the simple thermostat node is operating.
+Now let's see how the simple heater_voltage (Message) operates.
```
Comment (maksimdrachov) [inline docs/pages/demo.rst:190]: Correct? As I understand it: `heater_voltage` is a Message block (I keep wanting to say Message service, but I guess that would be a bit confusing) inside the `demo_app` node?

**Comment by maksimdrachov** [comment]
The last one is a useless change, but:

image

(Feel free to roll back)

File: docs/pages/demo.rst [hunk 2]
```diff
@@ -187,7 +310,7 @@ launch the following in a new terminal and leave it running (``y`` is a convenie
     export UAVCAN__UDP__IFACE=127.0.0.1
     y sub --with-metadata uavcan.node.heartbeat uavcan.diagnostic.record    # You should see heartbeats
 
-Now let's see how the simple thermostat node is operating.
+Now let's see how the simple heater_voltage (Message-publisher) operates.
```
Comment (maksimdrachov) [inline docs/pages/demo.rst:313]: Correct? As I understand it: `heater_voltage` is a _Message publisher_ inside the `demo_app` _node_?

**Comment by maksimdrachov** [comment]
This issue:

image

I've added `sphinxcontrib.contentui` in order to allow toggling the view of "Graph explanation" (see `docs/pages/demo.rst`). I think it's a bit clearer if the reader can toggle certain parts of the explanation that might be useful upon the first read, but migh otherwise be a bit too long to include in the default view of the page. 

So that leaves 2 options:
- remove this extension
- circumvent this check

File: demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl [hunk 3]
```diff
@@ -3,11 +3,11 @@
 
 PointXY.1.0[<64] points
 
-@extent 1024 * 8
+@extent 16 * 2 * 64 * 4
```
Comment (pavel-kirienko) [inline demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl:6]: I don't understand the purpose of this change. The original statement was correct -- the extent is set (arbitrarily) to 1 KiB. The size is not dependent on the fields contained in the definition, it is rather a top-level design requirement for the data type.

File: demo/demo_app.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 import pycyphal
 
 # DSDL files are automatically compiled by pycyphal import hook from sources pointed by CYPHAL_PATH env variable.
-import sirius_cyber_corp  # This is our vendor-specific root namespace. Custom data types.
+import arasaka_cyber_corp  # This is our vendor-specific root namespace. Custom data types.
```
Comment (pavel-kirienko) [inline demo/demo_app.py:13]: I prefer not to change this as there are references to this namespace on the forum that are better left valid. Also, I prefer my reference to the Hitchiker's Guide to the Galaxy to your reference to Cyberpunk.

File: docs/conf.py [hunk 5]
```diff
@@ -82,6 +82,8 @@
     "sphinx.ext.graphviz",
     "sphinx_computron",
     "ref_fixer_hack",
+    "sphinxcontrib.mermaid",
+    "sphinxcontrib.contentui",
```
Comment (pavel-kirienko) [inline docs/conf.py:86]: If it's not maintained, then it should not be used. I am not aware of alternatives and I think the value of this feature is comparatively low, so I would not expend extra effort on implementing it. We have much more important things to do.

File: docs/pages/demo.rst (hunk 2 above)
Comment (pavel-kirienko) [reply docs/pages/demo.rst:313]: It's called simply "publisher". I don't think this is a helpful replacement as we test not just a single publisher but rather the entire node here, which is a thermostat.

File: docs/pages/demo.rst [hunk 6]
```diff
@@ -242,7 +365,9 @@ For example, we can change the PID gains of the thermostat:
 
 ..  code-block:: sh
 
-    y r thermostat.pid.gains 2 0 0
+    y r 42 thermostat.pid.gains       # read current values
+    y r 42 thermostat.pid.gains 2 0 0 # write new values
+    y r 42 thermostat.pid.gains       # check values
```
Comment (pavel-kirienko) [inline docs/pages/demo.rst:370]: The last one is redundant because the new values are returned after the write call.

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]
We already use Graphviz for visualization purposes. Can we not draw the new diagrams in that instead of pulling in a yet another dependency?

https://edotor.net/

**Comment by pavel-kirienko** [comment]
@maksimdrachov nudge

File: demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl (hunk 3 above)
Comment (maksimdrachov) [reply demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl]: > Graphviz comment

```
digraph G {

    subgraph cluster {
        label = "42:org:opencyphal.pycyphal.demo.demo_app";
        node [shape=box]

        subgraph cluster_1 {
            label = "heartbeat_publisher";
            heartbeat_publisher_node[label="uavcan.node.Hearbeat.1.0", shape=trapezium, style=filled]
            uavcan_node_heartbeat[style=filled]
        }
        heartbeat_publisher_node -> uavcan_node_heartbeat
        

        subgraph cluster_2 {
            label = "temperature_setpoint";
            uavcan_si_sample_temperature_scalar_2345[style=filled]
            temperature_setpoint_node[label="uavcan.si.unit.temperature.Scalar_1", shape=invtrapezium, style=filled]
        }
        uavcan_si_sample_temperature_scalar_2345 -> temperature_setpoint_node

        subgraph cluster_3 {
            label = "temperature_measurement";
            uavcan_si_sample_temperature_scalar_2346[style=filled]
            temperature_measurement_node[label="uavcan.si.sample.temperature.Scalar_1", shape=invtrapezium, style=filled]
        }
        uavcan_si_sample_temperature_scalar_2346 -> temperature_measurement_node
        

        subgraph cluster_4 {
            label = "heater_voltage";
            heater_voltage_node[label="uavcan.si.unit.voltage.Scalar_1", shape=trapezium, style=filled]
            uavcan_si_unit_voltage_Scalar[label="2347:uavcan.si.unit.voltage.Scalar", style=filled]
        }
        heater_voltage_node -> uavcan_si_unit_voltage_Scalar

        subgraph cluster_5 {
            label = "least_squares";
            least_squares_service[label="sirius_cyber_corp.PerformLinearLeastSquaresFit_1", shape=hexagon, style=filled]
            sirius_cyber_corp_PerformLinearLeastSquaresFit_1_Request_123[label="123:sirius_cyber_corp.PerformLinearLeastSquaresFit_1.Request", style=filled]
            sirius_cyber_corp_PerformLinearLeastSquaresFit_1_Response_123[label="123:sirius_cyber_corp.PerformLinearLeastSquaresFit_1.Response", style=filled]
        }
        sirius_cyber_corp_PerformLinearLeastSquaresFit_1_Request_123 -> least_squares_service
        least_squares_service -> sirius_cyber_corp_PerformLinearLeastSquaresFit_1_Response_123
        

    }
    
}
```

How to enforce the order of the subgraphs?

Comment (maksimdrachov) [reply demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl]: > The size is not dependent on the fields contained in the definition, it is rather a top-level design requirement for the data type.

So, just to make sure I understand: 

_It can't it set to less than the total size of fields contained in the definition?_

From [Specification](https://opencyphal.org/specification/Cyphal_Specification.pdf):

> Extent — the minimum amount of memory, in bits, that shall be allocated to store the serialized representa-
tion of the type. The extent of any type is always greater than or equal the maximal value of its bit length set. It
is always a multiple of 8.

So chosen value should always be:

- greater/equal to 
  - if equal 
       - no fields can be added later (-> might as well use `sealed` in that case I guess?)
- multiple of 8

_It's not really arbitrary and does depend on the underlying total size?_

File: demo/demo_app.py (hunk 4 above)
Comment (maksimdrachov) [reply demo/demo_app.py]: Never finished reading that book. :)

Cyberpunk is a lot cooler but ok.

File: demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl (hunk 3 above)
Comment (pavel-kirienko) [reply demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl]: Yes, your understanding is correct. The extent value used here was chosen rather arbitrarily. You can ask the DSDL front-end to emit the actual size of a type by placing a statement like this after the last field:

```python
@print _offset_.max / 8  # Divide by 8 to obtain bytes (DSDL is bit-oriented)
```

`PointXY` is 4 bytes large and is sealed; 63 (sic! note `<` not `<=`) instances of `PointXY` take 252 bytes, plus one byte for the array length prefix, resulting in 253 bytes total.

You can also do this to obtain the list of all possible lengths at a given offset:

```python
@print _offset_ / 8
```

Be sure to feed `-v` to Nunavut to see the print outputs (actually it shouldn't require verbosity above the default value, perhaps this should be fixed in Nunavut):

```sh
nnvg -v --target-language c sirius_cyber_corp
```

File: docs/requirements.txt [hunk 7]
```diff
@@ -8,3 +8,5 @@
 sphinx           == 4.3.*
 sphinx_rtd_theme == 1.0.*
 sphinx-computron ~= 1.0
+sphinxcontrib-mermaid == 0.8.*
+sphinxcontrib-contentui == 0.2.*
```
Comment (pavel-kirienko) [inline docs/requirements.txt:12]: It would appear that these are no longer needed or am I misreading the code

**Review by pavel-kirienko — CHANGES_REQUESTED** [review]

File: demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl (hunk 3 above)
Comment (maksimdrachov) [reply demo/custom_data_types/arasaka_cyber_corp/PerformLinearLeastSquaresFit.1.0.dsdl]: At some point I will need to read the entire Specification. 👍

**Comment by maksimdrachov** [comment]
Feel free to merge, or keep open (since I might want to change some other small stuff around as well)

**Review by pavel-kirienko — APPROVED** [review]

**Comment by pavel-kirienko** [comment]
>I might want to change some other small stuff around as well

That's fine, let's take these changes in as they come in separate PRs.

**Pull request merged**